### ISSUE TITLE
kexec fails on azure

### ISSUE THREAD

**Issue by ltworf** [systemd/systemd#20072]
**systemd version the issue has been seen with**

237-3ubuntu10.47

**Used distribution**

ubuntu 18.04

**Linux kernel version used** (`uname -a`)

`Linux asd.com 5.4.0-73-generic #82~18.04.1-Ubuntu SMP Fri Apr 16 15:10:02 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux`

**CPU architecture issue was seen on**

amd64

**Expected behaviour you didn't see**

I expected a ≠ 0 return code

**Unexpected behaviour you saw**

kexec seemed to work but it wasn't actually supported on the platform.

**Steps to reproduce the problem**

1. Get a machine on azure
2. reboot with `systemctl kexec`

**Additional program output to the terminal or log subsystem illustrating the issue**

```
[    2.816028] zswap: loaded using pool lzo/zbud
[    2.821223] Key type ._fscrypt registered
[    2.825764] Key type .fscrypt registered
[    2.830746] Key type big_key registered
[    2.835569] Key type encrypted registered
[    2.840121] AppArmor: AppArmor sha1 policy hashing enabled
[    2.846950] ima: No TPM chip found, activating TPM-bypass!
[    2.853059] ima: Allocated hash algorithm: sha1
[    2.858198] ima: No architecture policies found
[    2.863130] evm: Initialising EVM extended attributes:
[    2.868985] evm: security.selinux
[    2.873100] evm: security.SMACK64
[    2.877064] evm: security.SMACK64EXEC
[    2.887169] evm: security.SMACK64TRANSMUTE
[    2.891907] evm: security.SMACK64MMAP
[    2.896247] evm: security.apparmor
[    2.900284] evm: security.ima
[    2.903966] evm: security.capability
[    2.908163] evm: HMAC attrs: 0x1
[    2.912376] PM:   Magic number: 9:922:730
[    2.918108] rtc_cmos 00:00: setting system clock to 2021-06-29T10:43:41 UTC (1624963421)
[    2.927735] Unstable clock detected, switching default tracing clock to "global"
[    2.927735] If you want to keep using the local clock, then add:
[    2.927735]   "trace_clock=local"
[    2.927735] on the kernel command line
[    2.952510] Freeing unused decrypted memory: 2040K
[    2.961967] Freeing unused kernel image memory: 2716K
[    2.967795] Write protecting the kernel read-only data: 22528k
[    2.974776] Freeing unused kernel image memory: 2008K
[    2.980640] Freeing unused kernel image memory: 1164K
[    2.995081] x86/mm: Checked W+X mappings: passed, no W+X pages found.
[    3.015522] x86/mm: Checking user space page tables
[    3.030663] x86/mm: Checked W+X mappings: passed, no W+X pages found.
[    3.038048] Run /init as init process
[    3.042717] Failed to execute /init (error -2)
[    3.048055] Run /sbin/init as init process
[    3.053536] Run /etc/init as init process
[    3.058709] Run /bin/init as init process
[    3.064801] Run /bin/sh as init process
[    3.074292] Kernel panic - not syncing: No working init found.  Try passing init= option to kernel. SeeLinux Documentation/admin-guide/init.rst for guidance.
[    3.078254] CPU: 0 PID: 1 Comm: swapper/0 Not tainted 5.4.0-77-generic #86~18.04.1-Ubuntu
[    3.078254] Hardware name: Microsoft Corporation Virtual Machine/Virtual Machine, BIOS 090008  12/07/2018
[    3.078254] Call Trace:
[    3.078254]  dump_stack+0x6d/0x8b
[    3.078254]  ? __noinstr_text_end+0x36f1/0x3730
[    3.078254]  panic+0xfe/0x2e0
[    3.078254]  ? rest_init+0xb0/0xb0
[    3.078254]  kernel_init+0xfa/0x110
[    3.078254]  ret_from_fork+0x35/0x40
[    3.078254] Kernel Offset: 0x1ee00000 from 0xffffffff81000000 (relocation range: 0xffffffff80000000-0xffffffffbfffffff)
[    3.078254] ---[ end Kernel panic - not syncing: No working init found.  Try passing init= option to kernel. See Linux Documentation/admin-guide/init.rst for guidance. ]---
```

Basically it seems that on azure, kexec is not supported: https://docs.microsoft.com/en-us/troubleshoot/windows-client/virtualization/cant-use-kdump-kexec-linux-virtual-machines-hyper-v

I'm currently using systemd-detect-virt to know if it's `microsoft` in which case I do a regular reboot, but I think `systemctl` should be failing to do the kexec thing, which i'd prefer rather that a kernel panic.

Not a cut and dry issue I guess, so I don't really have a precise resolution in mind other than failing.

**Comment by poettering** [comment]
the kexec fails because there's apparently no init system in the invoked image, see error messages. Of course kexec only works if if the loaded kernel/initrd stuff work fine. it's the distro's job to put that together.

the way i read the hyperv docs, kexec works, it's just some driver tht doesn't, but if you don't use that it works. 

Either way this is not a systemd issue as it appears, but between your distro and azure. Please ping those communities, they mighthelp you.
